gRPC bidi streaming testing in DevUI is broken

### Describe the bug

Bidirectional streaming in the DevUI test console is broken: 

```
2024-09-16 07:59:51,785 ERROR [io.qua.dev.run.com.JsonRpcRouter] (vert.x-eventloop-thread-2) Unable to invoke method io.quarkus.quarkus-grpc.streamService using JSON-RPC, request was: {
  "jsonrpc" : "2.0",
  "method" : "io.quarkus.quarkus-grpc.streamService",
  "params" : {
    "serviceName" : "test.Streaming",
    "methodName" : "pipe",
    "methodType" : "BIDI_STREAMING",
    "content" : "{\n  \"value\": \"dsaf\"\n}"
  },
  "id" : 11
}: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:118)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.devui.runtime.comms.JsonRpcRouter.routeToRuntimeSubscription(JsonRpcRouter.java:176)
	at io.quarkus.devui.runtime.comms.JsonRpcRouter.routeToRuntime(JsonRpcRouter.java:163)
	at io.quarkus.devui.runtime.comms.JsonRpcRouter.route(JsonRpcRouter.java:139)
	at io.quarkus.devui.runtime.comms.JsonRpcRouter.lambda$addSocket$0(JsonRpcRouter.java:105)
	at io.vertx.core.http.impl.WebSocketImplBase$FrameAggregator.handleTextFrame(WebSocketImplBase.java:634)
	at io.vertx.core.http.impl.WebSocketImplBase$FrameAggregator.handle(WebSocketImplBase.java:600)
	at io.vertx.core.http.impl.WebSocketImplBase$FrameAggregator.handle(WebSocketImplBase.java:589)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
	at io.vertx.core.http.impl.WebSocketImplBase.receiveFrame(WebSocketImplBase.java:534)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:279)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:157)
	at io.vertx.core.http.impl.WebSocketImplBase.handleFrame(WebSocketImplBase.java:475)
	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:313)
	at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:159)
	at io.vertx.core.http.impl.Http1xConnectionBase.handleWsFrame(Http1xConnectionBase.java:66)
	at io.vertx.core.http.impl.Http1xServerConnection.handleOther(Http1xServerConnection.java:191)
	at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:176)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:159)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:289)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1407)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalArgumentException: wrong number of arguments: 2 expected: 1
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.checkArgumentCount(DirectMethodHandleAccessor.java:329)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:101)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.grpc.runtime.devui.GrpcJsonRPCService.streamService(GrpcJsonRPCService.java:148)
	at io.quarkus.grpc.runtime.devui.GrpcJsonRPCService_ClientProxy.streamService(Unknown Source)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	... 45 more
```

### Expected behavior

It works.

### Actual behavior

Fails with the exception.

### How to Reproduce?

Reproducer - https://github.com/xstefank/quarkus-reproducers/tree/main/grpc-bidi-streaming

Start dev mode.
Access the gRPC card in DevUI.
Try to pass some value into the bidi_streaming pipe method of test.Streaming.
It fails with the above exception.

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/extensions/grpc/deployment/src/main/resources/dev-ui/qwc-grpc-services.js
+++ b/extensions/grpc/deployment/src/main/resources/dev-ui/qwc-grpc-services.js
@@ -233,20 +233,22 @@ export class QwcGrpcServices extends observeState(QwcHotReloadElement) {
     
     _renderCommandButtons(service, method){
         if(this._streamsMap.size >=0){
-            if(method.type == 'UNARY'){
-                return html`<vaadin-button theme="secondary error" @mousedown=${() => this._clear(service.name, method)} @mouseup=${() => this._default(service.name, method)}>Reset</vaadin-button>
+            if(method.type == 'UNARY' || method.type == 'SERVER_STREAMING'){
+                return html`<vaadin-button theme="secondary error" @click=${() => this._default(service.name, method)}>Reset</vaadin-button>
                             <vaadin-button theme="secondary success" @click=${() => this._test(service, method)}>Send</vaadin-button>`;
             }else if(this._isRunning(service.name, method)){
-                return html`<div class="streamButtons"><vaadin-button theme="secondary error" @click=${() => this._test(service, method)}>Cancel stream</vaadin-button>
-                            <vaadin-progress-bar class="progress-short" indeterminate></vaadin-progress-bar></div>`;
+                return html`<vaadin-button theme="secondary error" @click=${() => this._default(service.name, method)}>Reset</vaadin-button>
+                            <vaadin-button theme="secondary success" @click=${() => this._test(service, method)}>Send</vaadin-button>
+                            <vaadin-button theme="secondary error" @click=${() => this._disconnect(service, method)}>Disconnect</vaadin-button>
+                            <vaadin-progress-bar class="progress-short" indeterminate></vaadin-progress-bar>`;
             }else {
-                return html`<vaadin-button theme="secondary success" @click=${() => this._test(service, method)}>Start stream</vaadin-button>`;
+                return html`<vaadin-button theme="secondary success" @click=${() => this._test(service, method)}>Send</vaadin-button>`;
             }
         }
     }
     
     _keypress(e, service, method){
-        if(method.type == 'UNARY' || !this._isRunning(service.name, method)){
+        if(method.type == 'UNARY' || method.type == 'SERVER_STREAMING' || !this._isRunning(service.name, method)){
             if ((e.keyCode == 10 || e.keyCode == 13) && e.ctrlKey){ // ctlr-enter
                 this._test(service, method);
             }
@@ -268,45 +270,74 @@ export class QwcGrpcServices extends observeState(QwcHotReloadElement) {
     }
     
     _default(serviceName, method){
+        let requestTextArea = this._requestTextArea(serviceName, method);
+        requestTextArea.content = '';
         let pv = JSON.parse(method.prototype);
-        this._requestTextArea(serviceName, method).populatePrettyJson(JSON.stringify(pv));
+        let prettyJson = JSON.stringify(pv, null, 2);
+        requestTextArea.populatePrettyJson(prettyJson);
     }
     
     _test(service, method){
-        let textArea = this._requestTextArea(service.name, method);
-        let content = textArea.getAttribute('value');
+        let requestTextArea = this._requestTextArea(service.name, method);
+        let content = requestTextArea.getAttribute('value');
+        let id = this._id(service.name, method);
+        let responseTextArea = this._responseTextArea(service.name, method);
         if(method.type == 'UNARY'){
             this.jsonRpc.testService({
+                id: id,
                 serviceName: service.name,
                 methodName: method.bareMethodName,
                 methodType: method.type,
                 content: content
             }).then(jsonRpcResponse => {
-                const jsonObject = JSON.parse(jsonRpcResponse.result);
-                const prettyJson = JSON.stringify(jsonObject, null, 2);
-                this._responseTextArea(service.name, method).populatePrettyJson(prettyJson);
+                this._responseTextArea(service.name, method).populatePrettyJson(this._prettyJson(jsonRpcResponse.result));
             });
         }else{
-            let id = this._id(service.name, method);
             if(this._isRunning(service.name, method)){
-                this._streamsMap.get(id).cancel();
-                this._streamsMap.delete(id);
-                this._clear(service.name, method);
-                this._default(service.name, method);
-            }else{
-                let cancelable = this.jsonRpc.streamService({
+                this.jsonRpc.streamService({
+                    id: id,
                     serviceName: service.name,
                     methodName: method.bareMethodName,
-                    methodType: method.type,
+                    isRunning: true,
+                    content: content
+                });
+                // this._streamsMap.get(id).cancel();
+                // this._streamsMap.delete(id);
+                // this._clear(service.name, method);
+                // this._default(service.name, method);
+            }else{
+                // starting a new stream, clear the response area
+                responseTextArea.content = null;
+                let cancelable = this.jsonRpc.streamService({
+                    id: id,
+                    serviceName: service.name,
+                    methodName: method.bareMethodName,
+                    isRunning: false,
                     content: content
                 }).onNext(jsonRpcResponse => {
-                    this._responseTextArea(service.name, method).populatePrettyJson(jsonRpcResponse.result);
+                    if (responseTextArea.content == null) {
+                        responseTextArea.populatePrettyJson(this._prettyJson(jsonRpcResponse.result));
+                    } else {
+                        responseTextArea.populatePrettyJson(responseTextArea.content + '\n' + this._prettyJson(jsonRpcResponse.result));
+                    }
                 });
-                this._streamsMap.set(id, cancelable);
+                if (method.type == 'BIDI_STREAMING' || method.type == 'CLIENT_STREAMING') {
+                    this._streamsMap.set(id, cancelable);
+                }
             }
             this._testerButtons = this._renderCommandButtons(service, method);
             this._forceUpdate();
         }
+    }
+
+    _disconnect(service, method){
+        let id = this._id(service.name, method);
+        this.jsonRpc.disconnectService({
+            id: id,
+        });
+        this._streamsMap.delete(id);
+        this._testerButtons = this._renderCommandButtons(service, method);
+        this._forceUpdate();
     }
     
     _forceUpdate(){
@@ -331,6 +362,10 @@ export class QwcGrpcServices extends observeState(QwcHotReloadElement) {
     
     _responseId(serviceName, method){
         return serviceName + '/' + method.bareMethodName + '_response';
+    }
+
+    _prettyJson(content){
+        return JSON.stringify(JSON.parse(content), null, 2);
     }
 }
 customElements.define('qwc-grpc-services', QwcGrpcServices);

--- a/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/devui/GrpcJsonRPCService.java
+++ b/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/devui/GrpcJsonRPCService.java
@@ -18,6 +18,7 @@ import com.google.protobuf.MessageOrBuilder;
 import com.google.protobuf.util.JsonFormat;
 
 import io.grpc.Channel;
+import io.grpc.ManagedChannel;
 import io.grpc.MethodDescriptor;
 import io.grpc.ServiceDescriptor;
 import io.grpc.netty.NettyChannelBuilder;
@@ -45,6 +46,7 @@ public class GrpcJsonRPCService {
     private static final Logger LOG = Logger.getLogger(GrpcJsonRPCService.class);
 
     private Map<String, GrpcServiceClassInfo> grpcServiceClassInfos;
+    private Map<String, StreamObserver<Message>> callsInProgress;
 
     @Inject
     HttpConfiguration httpConfiguration;
@@ -72,6 +74,7 @@ public class GrpcJsonRPCService {
             this.ssl = isTLSConfigured(httpConfiguration.ssl.certificate);
         }
         this.grpcServiceClassInfos = getGrpcServiceClassInfos();
+        this.callsInProgress = new HashMap<>();
     }
 
     private boolean isTLSConfigured(CertificateConfig certificate) {
@@ -107,25 +110,27 @@ public class GrpcJsonRPCService {
         return services;
     }
 
-    public Uni<String> testService(String serviceName, String methodName, String methodType, String content) {
+    public Uni<String> testService(String id, String serviceName, String methodName, String content) {
         try {
-            return streamService(serviceName, methodName, methodType, content).toUni();
+            return streamService(id, serviceName, methodName, false, content).toUni();
         } catch (Throwable t) {
             return Uni.createFrom().item(error(t.getMessage()).encodePrettily());
         }
     }
 
-    public Multi<String> streamService(String serviceName, String methodName, String methodType, String content)
+    public Multi<String> streamService(String id, String serviceName, String methodName, boolean isRunning,
+            String content)
             throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InvalidProtocolBufferException {
         if (content == null) {
-            return Multi.createFrom().item(error("Invalid messsge").encodePrettily());
+            return Multi.createFrom().item(error("Invalid message").encodePrettily());
         }
 
         BroadcastProcessor<String> streamEvent = BroadcastProcessor.create();
 
         GrpcServiceClassInfo info = this.grpcServiceClassInfos.get(serviceName);
 
-        Object grpcStub = createStub(info.grpcServiceClass, host, port);
+        ManagedChannel channel = getChannel(host, port);
+        Object grpcStub = createStub(info.grpcServiceClass, channel);
 
         ServiceDescriptor serviceDescriptor = info.serviceDescriptor;
 
@@ -134,20 +139,50 @@ public class GrpcJsonRPCService {
         MethodDescriptor.PrototypeMarshaller<?> protoMarshaller = (MethodDescriptor.PrototypeMarshaller<?>) requestMarshaller;
         Class<?> requestType = protoMarshaller.getMessagePrototype().getClass();
 
+        Message message = createMessage(content, requestType);
+
+        if (isRunning) {
+            // we are already connected with this gRPC endpoint, just send the message
+            callsInProgress.get(id).onNext(message);
+        } else {
+            // Invoke the stub method and format the response as JSON
+            StreamObserver<?> responseObserver = new TestObserver<>(streamEvent);
+            StreamObserver<Message> incomingStream;
+
+            final Method stubMethod = getStubMethod(grpcStub, methodDescriptor.getBareMethodName());
+
+            if (stubMethod.getParameterCount() == 1 && stubMethod.getReturnType() == StreamObserver.class) {
+                // returned StreamObserver consumes incoming messages
+                //noinspection unchecked
+                incomingStream = (StreamObserver<Message>) stubMethod.invoke(grpcStub, responseObserver);
+                callsInProgress.put(id, incomingStream);
+                // will be streamed continuously
+                incomingStream.onNext(message);
+            } else {
+                // incoming message should be passed as the first parameter of the invocation
+                stubMethod.invoke(grpcStub, message, responseObserver);
+            }
+        }
+
+        channel.shutdown();
+        return streamEvent;
+    }
+
+    private static Message createMessage(String content, Class<?> requestType)
+            throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InvalidProtocolBufferException {
         // Create a new builder for the request message, e.g. HelloRequest.newBuilder()
         Method newBuilderMethod = requestType.getDeclaredMethod("newBuilder");
         Message.Builder builder = (Message.Builder) newBuilderMethod.invoke(null);
 
         // Use the test data to build the request object
         JsonFormat.parser().merge(content, builder);
-        Message message = builder.build();
+        return builder.build();
+    }
 
-        StreamObserver<?> responseObserver = new TestObserver<Object>(streamEvent);
-
-        final Method stubMethod = getStubMethod(grpcStub, methodDescriptor.getBareMethodName());
-        stubMethod.invoke(grpcStub, message, responseObserver);
-
-        return streamEvent;
+    public Uni<Void> disconnectService(String id) {
+        callsInProgress.get(id).onCompleted();
+        callsInProgress.remove(id);
+        return Uni.createFrom().voidItem();
     }
 
     private Map<String, GrpcJsonRPCService.GrpcServiceClassInfo> getGrpcServiceClassInfos() {
@@ -220,17 +255,17 @@ public class GrpcJsonRPCService {
         return null;
     }
 
-    private Object createStub(Class<?> grpcServiceClass, String host, int port) {
+    private Object createStub(Class<?> grpcServiceClass, Channel channel) {
         try {
             Method stubFactoryMethod = grpcServiceClass.getDeclaredMethod("newStub", Channel.class);
-            return stubFactoryMethod.invoke(null, getChannel(host, port));
+            return stubFactoryMethod.invoke(null, channel);
         } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
             LOG.warnf("Could not create stub for %s - " + e.getMessage(), grpcServiceClass);
             return null;
         }
     }
 
-    private Channel getChannel(String host, int port) {
+    private ManagedChannel getChannel(String host, int port) {
         return NettyChannelBuilder
                 .forAddress(host, port)
                 .usePlaintext()

